core/src/main/java/org/apache/iceberg/PartitionsTable.java

```diff
@@ -220,21 +251,53 @@ Iterable<Partition> all() {
 
   static class Partition {
     private final StructLike key;
-    private long recordCount;
-    private int fileCount;
     private int specId;
+    private long dataRecordCount;
+    private int dataFileCount;
+
+    private final Set<DeleteFile> equalityDeleteFiles;
+    private final Set<DeleteFile> positionDeleteFiles;
```

Comment:
Can you double check DeleteFile equal/hashcode are working correctly, if conflict?  Should we rather do something safer, like paths?

Reply:
https://github.com/apache/iceberg/blob/76c1df0cfde27167d31c112c81daa2d4fcc96686/api/src/main/java/org/apache/iceberg/RewriteFiles.java#L71

`Set<DeleteFile>` was already used in the existing code, as seen above. 
But there are no equals() and hashcode() impl for these classes. Which means it is using the default ones. 

Note that `Set<DataFile>` also exist in the code and it also doesn't have equal/hashcode impl. 

I am ok with using Paths. But just wondering why it doesn't exist and maybe we can handle these in a follow-up PR instead of this one.

Reply:
So I am not so confident, that if for example you have two FileScanTask, returning deletes() that are the same, will the two DeleteFile object be different or not?  Java default equals() is instance equality, isnt it?  And even if it does work, will break if we implement those at some point.  Hence the suggestion to use something with established hashCode/equals.

Reply:
The same delete file object from the context is reused while making the fileScan here. 
https://github.com/apache/iceberg/blob/d42d1e89c0616c203f7ad29f002811ddd440e14f/core/src/main/java/org/apache/iceberg/ManifestGroup.java#L351

hence, I believe the default equals() is enough. 

I don't disagree about having the equals and hashcode for `DataFiles` and `DeleteFiles`. But code is widely using the Set<DataFile> and Set<DeleteFile> already. So, if we are adding it. It should be in a separate PR/discussion. 

Let us see what others think on this. 
cc: @RussellSpitzer, @jackye1995

Reply:
Yea it may work due to cache inside DeleteIndex, but I don't know.  I see Set<DeleteFile> but it seems those case are where they are limited to not conflict with each other except instance equality.  I would also be interested in the best way here.

Reply:
I'd be more worried about the memory usage here, seems like we have to keep the entire set of all delete files in memory (for all partitions) while we are building this table?

Reply:
done.